File: Day25/Java/Main.java
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.List;

public class Main {

	public static void main(String[] args) throws IOException {
		File input = new File("input").getAbsoluteFile();
		if (!input.exists()) {
			System.out.println(input.getPath() + " does not exist!");
			System.out.println("trying " + input.getPath() + ".txt instead.");
			input = new File(input.getPath() + ".txt");
			if (!input.exists()) {
				System.err.println("neither " + input.getPath().substring(0, input.getPath().length() - 4) + ",");
				System.err.println("nor " + input.getPath() + " exists!");
				return;
			}
		}

		List<String> lines = Files.readAllLines(input.toPath());
		final int subjectNumber = 7;
		final int key1 = Integer.parseInt(lines.get(0));
		final int key2 = Integer.parseInt(lines.get(1));

		final long encryptionKey = getKey(key2, getLoopSize(key1, subjectNumber));
		System.out.printf("The encryption key is %d.", encryptionKey);
	}

	public static long getKey(int subjectNumber, int loopSize) {
		long value = 1;
		for (int i = 0; i < loopSize; i++) {
			value = value * subjectNumber % 20201227;
		}
		return value;
	}

	public static int getLoopSize(int key, int subjectNumber) {
		int loopSize = 0;
		while (key != 1) {
			double newKey = key / (double) subjectNumber;
			while (newKey % 1 != 0) {
				key += 20201227;
				newKey = key / (double) subjectNumber;
			}
			key = (int) newKey;
			loopSize++;
		}
		return loopSize;
	}
}

File: Day25/Python/Main.py
import os

def getLoopSize(key, subjectNumber):
    loopSize = 0
    while key != 1:
        newKey = key / subjectNumber
        while newKey % 1 != 0:
            key += 20201227
            newKey = key / subjectNumber

        key = newKey
        loopSize += 1

    return loopSize

def getKey(subjectNumber, loopSize):
    value = 1
    for _ in range(loopSize):
        value = value * subjectNumber % 20201227
    return value

def main():
    input = os.path.dirname(os.getcwd())
    input = os.path.join(input, "input")
    if not os.path.exists(input):
        print(input + " does not exist!")
        input = os.path.join(os.path.dirname(input), "input.txt")
        print(f"trying {input} instead.")
        if not os.path.exists(input):
            print(input + " does not exist!")
            return;

    file = open(input, 'r')
    subjectNumber = 7
    key1 = int(file.readline())
    key2 = int(file.readline())
    file.close()

    encryptionKey = getKey(key2, getLoopSize(key1, subjectNumber))
    print(f"The encryption key is {encryptionKey}.")

if __name__ == '__main__':
    main()
